rust/codelist-builder-rs/src/snomed_usage_data.rs

```diff
@@ -0,0 +1,391 @@
+        assert!(!entries[9].active_at_start);
+        assert!(!entries[9].active_at_end);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_parse_from_string_empty_data() -> Result<(), CodeListBuilderError> {
+        let test_data = "SNOMED_Concept_ID	Description	Usage	Active_at_Start	Active_at_End";
+        let entries = SnomedUsageData::parse_from_string(test_data)?;
+        assert_eq!(entries.len(), 0);
+        Ok(())
+    }
+
+    #[test]
+    fn test_parse_from_string_column_count_too_small() -> Result<(), CodeListBuilderError> {
+        let test_data = "SNOMED_Concept_ID	Description	Usage	Active_at_Start
+279991000000102	Short message service text message sent to patient (procedure)	122292090	1";
+        let error = SnomedUsageData::parse_from_string(test_data).unwrap_err();
+        let error_string = error.to_string();
+        assert_eq!(
+            &error_string,
+            "Invalid usage data: Invalid number of columns in record (4) at row 1"
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_parse_from_string_column_count_too_big() -> Result<(), CodeListBuilderError> {
+        let test_data =
+            "SNOMED_Concept_ID	Description	Usage	Active_at_Start	Active_at_End	Active_at_End
+279991000000102	Short message service text message sent to patient (procedure)	122292090	1	1	1";
+        let error = SnomedUsageData::parse_from_string(test_data).unwrap_err();
+        let error_string = error.to_string();
+        assert_eq!(
+            &error_string,
+            "Invalid usage data: Invalid number of columns in record (6) at row 1"
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_parse_from_string_unequal_column_count() -> Result<(), CodeListBuilderError> {
+        let test_data = "SNOMED_Concept_ID	Description	Usage	Active_at_Start	Active_at_End
+279991000000102	Short message service text message sent to patient (procedure)	122292090	1
+163030003	On examination - Systolic blood pressure reading (finding)	59227180	1
+163031004	On examination - Diastolic blood pressure reading (finding)	59184050	1	1
+163020007	On examination - blood pressure reading (finding)	37837700	1	1
+1000731000000107	Serum creatinine level (observable entity)	33211250	1
+1000661000000107	Serum sodium level (observable entity)	31630420	1	1";
+        let error = SnomedUsageData::parse_from_string(test_data).unwrap_err();
+        let error_string = error.to_string();
+        assert!(error_string.contains("CSV error:"));
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_download_usage_from_url() -> Result<(), CodeListBuilderError> {
+        let mock_server = MockServer::start().await;
+        let usage_year = UsageYear::Y2020_21;
+
+        let test_data = LONG_TEST_DATA;
+
+        Mock::given(method("GET"))
+            .and(path(usage_year.path()))
+            .respond_with(ResponseTemplate::new(200).set_body_string(test_data))
+            .mount(&mock_server)
+            .await;
+
```

Comment:
This is cool to see. I hadn't come across this crate before. 

That being said I am not sure we really need to test the functionality of what is essentially a `reqwest` call. It might be better to concentrate on calls that fail and if the error message that gets passed back is understandable. has the url changed? is the file corrupted? is it too big to be downloaded? etc. 

for now, lets keep this test in as it is useful documentation anyway.